This is the R Notebook for the R assignment. 

General Notes: 
When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).


#Part One: Data Inspection

First I downloaded the files needed for the assignment:

```{r}
fang <- read.table("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-fall2017/master/UNIX_Assignment/fang_et_al_genotypes.txt", sep = "\t", header = TRUE, na.strings = "?/?", stringsAsFactors = FALSE)
```
```{r}
snp_pos <- read.table("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2017/master/UNIX_Assignment/snp_position.txt", sep = "\t", header = TRUE, na.strings = "?/?", stringsAsFactors = FALSE)
```

This allowed me to download the file, indicate that the files were tab delimited, there are headers in the first row, replaced any ?/? with NA, and prevented the automatic conversion of character strings to factors


Next, I viewed the data to take a quick look that all the data had downloaded correctly:

```{r}
View(fang)
View(snp_pos)
```

I then examine the structure and dimension of the fang et al files:
```{r}
typeof(fang)
class(fang)
ncol(fang)
nrow(fang)

```
I examine the structure and dimensions of the snp position files:

```{r}
typeof(snp_pos)
class(snp_pos)
ncol(snp_pos)
nrow(snp_pos)
```
#Part One: Data Processing

I filter out the maize files from the fang et al file, first by creating a vector with the maize group IDs, then pulling the rows with those IDs into a new file. I then view the files to verify:

```{r}
maize_snps <- filter(fang, Group== "ZMMIL" | Group=="ZMMLR" | Group== "ZMMMR")
View(maize_snps)
```

I also filter out the teosinte files with the same process. First by creating a vector with the group IDs, then pulling the rows with those IDs into a new file, followed by a view to verify:
```{r}
teosinte_snps <- filter(fang, Group== "ZMPBA" | Group== "ZMPIL" | Group== "ZMPJA")
View(maize_snps)
```

Then I transposed the genotype data in order to join the two dataframes:
```{r}
transp_maize_snps <- t(maize_snps)
transp_teo_snps <- t(teosinte_snps)
```
I then review the transposed data:
```{r}
View(transp_maize_snps)
View(transp_teo_snps)
typeof(transp_maize_snps)
class(transp_maize_snps)
typeof(transp_teo_snps)
class(transp_teo_snps)
```

The transposed data vector was converted to characters/matrix, when as listed above, it should be list/data.frame. Therefore, I converted the transposed data into a dataframe, with a list type:

```{r}
df_trns_maize_snps <- as.data.frame(transp_maize_snps, stringsAsFactors = FALSE)
df_trns_teo_snps <- as.data.frame(transp_teo_snps, stringsAsFactors = FALSE)
```
I then reviewed the dataframe'd transposed data:

```{r}
View(df_trns_maize_snps)
View(df_trns_teo_snps)

typeof(df_trns_maize_snps)
class(df_trns_maize_snps)
typeof(df_trns_teo_snps)
class(df_trns_teo_snps)
```
With the values in the correct vector, I then merged the data frames together

```{r}
merged_mz_data <- merge.data.frame(snp_pos, df_trns_maize_snps, by.x = 1, by.y = 0)
merged_teo_data <- merge.data.frame(snp_pos, df_trns_teo_snps, by.x = 1, by.y = 0)
View(merged_mz_data)
View(merged_teo_data)
ncol(merged_mz_data)
ncol(merged_teo_data)
typeof(merged_mz_data)
class(merged_mz_data)
```


Now that the data has been merged, the files need to be ordered in the following order:

Column 1: SNP_ID  (currently column 1)
Column 2: Chromosome (currently column 3)
Column 3: Position (currently column 4)
Subsq col: all other genotype data (columns )

So I reordered the columsn in the following order:

maize: 1, 3, 4, 16-1588
teosinte: 1, 3, 4, 16-990

Omitting columns 2 and 5-15 as non-genotype data and thus not needed.

```{r}
ord_mz_data <- merged_mz_data[,c(1,3,4,16:1588)]
ord_teo_data <- merged_teo_data[, c(1,3,4,16:990)]
View(ord_mz_data)
View(ord_teo_data)
```
```{r}
typeof(ord_mz_data)
class(ord_mz_data)
```
I reviewed the type and class of both Position and Chromosome and found they were both Character:

```{r}
typeof(ord_mz_data$Position)
class(ord_mz_data$Position)
typeof(ord_mz_data$Chromosome)
class(ord_mz_data$Chromosome)
```
Therefore, I needed to convert both to factors:

```{r}
ord_mz_data$Chromosome <- as.factor(ord_mz_data$Chromosome)
ord_mz_data$Position <- as.factor(ord_mz_data$Position)
ord_teo_data$Chromosome <- as.factor(ord_teo_data$Chromosome)
ord_teo_data$Position <- as.factor(ord_teo_data$Position)
class(ord_mz_data$Position)
class(ord_mz_data$Chromosome)
class(ord_teo_data$Position)
class(ord_teo_data$Chromosome)

```

Now that they are in factors, I will make four files, two with question marks for the missing files and two with dashes:

```{r}
ord_mz_data_ques <- replace(ord_mz_data, is.na(ord_mz_data), "?")
ord_teo_data_ques <- replace(ord_teo_data, is.na(ord_teo_data), "?")
ord_mz_data_dash <- replace(ord_mz_data, is.na(ord_mz_data), "-")
ord_teo_data_dash <- replace(ord_teo_data, is.na(ord_teo_data),"-")
```

So now that they are in a format that can be ordered, I'm going to order them in increasing and decreasing position.

```{r}
ord_maize_inc <- ord_mz_data_ques[order(ord_mz_data_ques$Position, decreasing=FALSE),]
ord_maize_dec <- ord_mz_data_dash[order(ord_mz_data_dash$Position, decreasing=TRUE),]
ord_teo_inc <- ord_teo_data_ques[order(ord_teo_data_ques$Position, decreasing=FALSE),]
ord_teo_dec <- ord_teo_data_dash[order(ord_teo_data_dash$Position, decreasing=TRUE),]
```
Next can subset them into the appropriate chromosomes, first with increasing position and ? for missing data:

```{r}
Chr1_maize_inc <- subset(ord_maize_inc, Chromosome=="1")
Chr2_maize_inc <- subset(ord_maize_inc, Chromosome=="2")
Chr3_maize_inc <- subset(ord_maize_inc, Chromosome=="3")
Chr4_maize_inc <- subset(ord_maize_inc, Chromosome=="4")
Chr5_maize_inc <- subset(ord_maize_inc, Chromosome=="5")
Chr6_maize_inc <- subset(ord_maize_inc, Chromosome=="6")
Chr7_maize_inc <- subset(ord_maize_inc, Chromosome=="7")
Chr8_maize_inc <- subset(ord_maize_inc, Chromosome=="8")
Chr9_maize_inc <- subset(ord_maize_inc, Chromosome=="9")
Chr10_maize_inc <- subset(ord_maize_inc, Chromosome=="10")

Chr1_teo_inc <- subset(ord_teo_inc, Chromosome=="1")
Chr2_teo_inc <- subset(ord_teo_inc, Chromosome=="2")
Chr3_teo_inc <- subset(ord_teo_inc, Chromosome=="3")
Chr4_teo_inc <- subset(ord_teo_inc, Chromosome=="4")
Chr5_teo_inc <- subset(ord_teo_inc, Chromosome=="5")
Chr6_teo_inc <- subset(ord_teo_inc, Chromosome=="6")
Chr7_teo_inc <- subset(ord_teo_inc, Chromosome=="7")
Chr8_teo_inc <- subset(ord_teo_inc, Chromosome=="8")
Chr9_teo_inc <- subset(ord_teo_inc, Chromosome=="9")
Chr10_teo_inc <- subset(ord_teo_inc, Chromosome=="10")
```

Finally, I can subset them with decreasing position and - for missing data:

```{r}
Chr1_maize_dec <- subset(ord_maize_dec, Chromosome=="1")
Chr2_maize_dec <- subset(ord_maize_dec, Chromosome=="2")
Chr3_maize_dec <- subset(ord_maize_dec, Chromosome=="3")
Chr4_maize_dec <- subset(ord_maize_dec, Chromosome=="4")
Chr5_maize_dec <- subset(ord_maize_dec, Chromosome=="5")
Chr6_maize_dec <- subset(ord_maize_dec, Chromosome=="6")
Chr7_maize_dec <- subset(ord_maize_dec, Chromosome=="7")
Chr8_maize_dec <- subset(ord_maize_dec, Chromosome=="8")
Chr9_maize_dec <- subset(ord_maize_dec, Chromosome=="9")
Chr10_maize_dec <- subset(ord_maize_dec, Chromosome=="10")

Chr1_teo_dec <- subset(ord_teo_dec, Chromosome=="1")
Chr2_teo_dec <- subset(ord_teo_dec, Chromosome=="2")
Chr3_teo_dec <- subset(ord_teo_dec, Chromosome=="3")
Chr4_teo_dec <- subset(ord_teo_dec, Chromosome=="4")
Chr5_teo_dec <- subset(ord_teo_dec, Chromosome=="5")
Chr6_teo_dec <- subset(ord_teo_dec, Chromosome=="6")
Chr7_teo_dec <- subset(ord_teo_dec, Chromosome=="7")
Chr8_teo_dec <- subset(ord_teo_dec, Chromosome=="8")
Chr9_teo_dec <- subset(ord_teo_dec, Chromosome=="9")
Chr10_teo_dec <- subset(ord_teo_dec, Chromosome=="10")

```
To record the actual files to GitHub:

```{r}

write.csv(Chr1_maize_inc, file = "Chr1_maize_inc.csv")
write.csv(Chr2_maize_inc, file = "Chr2_maize_inc.csv")
write.csv(Chr3_maize_inc, file = "Chr3_maize_inc.csv")
write.csv(Chr4_maize_inc, file = "Chr4_maize_inc.csv")
write.csv(Chr5_maize_inc, file = "Chr5_maize_inc.csv")
write.csv(Chr6_maize_inc, file = "Chr6_maize_inc.csv")
write.csv(Chr7_maize_inc, file = "Chr7_maize_inc.csv")
write.csv(Chr8_maize_inc, file = "Chr8_maize_inc.csv")
write.csv(Chr9_maize_inc, file = "Chr9_maize_inc.csv") 
write.csv(Chr10_maize_inc, file = "Chr10_maize_inc.csv")

write.csv(Chr1_teo_inc, file = "Chr1_teo_inc.csv")
write.csv(Chr2_teo_inc, file = "Chr2_teo_inc.csv")
write.csv(Chr3_teo_inc, file = "Chr3_teo_inc.csv")
write.csv(Chr4_teo_inc, file = "Chr4_teo_inc.csv")
write.csv(Chr5_teo_inc, file = "Chr5_teo_inc.csv")
write.csv(Chr6_teo_inc, file = "Chr6_teo_inc.csv")
write.csv(Chr7_teo_inc, file = "Chr7_teo_inc.csv")
write.csv(Chr8_teo_inc, file = "Chr8_teo_inc.csv")
write.csv(Chr9_teo_inc, file = "Chr9_teo_inc.csv")
write.csv(Chr10_teo_inc, file = "Chr10_teo_inc.csv")

write.csv(Chr1_maize_dec, file = "Chr1_maize_dec.csv")
write.csv(Chr2_maize_dec, file = "Chr2_maize_dec.csv")
write.csv(Chr3_maize_dec, file = "Chr3_maize_dec.csv")
write.csv(Chr4_maize_dec, file = "Chr4_maize_dec.csv")
write.csv(Chr5_maize_dec, file = "Chr5_maize_dec.csv")
write.csv(Chr6_maize_dec, file = "Chr6_maize_dec.csv")
write.csv(Chr7_maize_dec, file = "Chr7_maize_dec.csv")
write.csv(Chr8_maize_dec, file = "Chr8_maize_dec.csv")
write.csv(Chr9_maize_dec, file = "Chr9_maize_dec.csv")
write.csv(Chr10_maize_dec, file = "Chr10_maize_dec.csv")

write.csv(Chr1_teo_dec, file = "Chr1_teo_dec.csv")
write.csv(Chr2_teo_dec, file = "Chr2_teo_dec.csv")
write.csv(Chr3_teo_dec, file = "Chr3_teo_dec.csv")
write.csv(Chr4_teo_dec, file = "Chr4_teo_dec.csv")
write.csv(Chr5_teo_dec, file = "Chr5_teo_dec.csv")
write.csv(Chr6_teo_dec, file = "Chr6_teo_dec.csv")
write.csv(Chr7_teo_dec, file = "Chr7_teo_dec.csv")
write.csv(Chr8_teo_dec, file = "Chr8_teo_dec.csv")
write.csv(Chr9_teo_dec, file = "Chr9_teo_dec.csv")
write.csv(Chr10_teo_dec, file = "Chr10_teo_dec.csv")
```

#Part Two: SNPs per Chromosome

To visualize the data, I sort out unneeded columns and used melt to tidy up the data:

```{r}
ncol(ord_mz_data)
ncol(ord_teo_data)
mz_snp_per_chrom <- ord_mz_data[, c(2,4:1576)]
teo_snp_per_chrom <- ord_teo_data[, c(2,4:978)]
melt_mz_snp_per_chrom <- melt(mz_snp_per_chrom, id.vars = "Chromosome", variable.name = "SNP column", na.rm = FALSE, value.name = "SNPs", factorAsStrings = TRUE)
melt_teo_snp_per_chrom <- melt(teo_snp_per_chrom, id.vars = "Chromosome", variable.name = "SNP column", na.rm = FALSE, value.name = "SNPs", factorAsString = TRUE)

```
Now I visualize the data by plot. First for the maize data: 


```{r}
tally_mz_SPC <- melt_mz_snp_per_chrom %>% count(Chromosome)
Maize_Chromsome_No <- tally_mz_SPC$Chromosome
Maize_Total_SNPs <- tally_mz_SPC$n
plot(Maize_Chromsome_No,Maize_Total_SNPs, xlab="Maize Chromosome_No", ylab = "Total SNPs")
```
**Plot was working previously, as can be see in the plot imgs in GitHub. Chromosome column still present.**

Based on the plot above, for the maize, chromosome 1 contributed the most of these SNPs

Next I visualize the plot for teosinte:

```{r}
tally_teo_SPC <- melt_teo_snp_per_chrom %>% count(Chromosome)
Teosinte_Chromsome_No <- tally_teo_SPC$Chromosome
Teosinte_Total_SNPs <- tally_teo_SPC$n
plot(Teosinte_Chromsome_No, Teosinte_Total_SNPs,  xlab="Teosinte Chromosome No", ylab="Total SNPs")
```
**Same error as above, please see saved plot in GitHub**

Based on the plot above, for the teosinte, chromosome 1 contributed the most of these SNPs.


#Part Two: Missing Data and Amt of Heterozygosity

Next, I create, sort and melt the maize and teosinte snps files, in order to tidy up for additional visualizations:
```{r}
zygos_mz <- maize_snps[, c(1,3,5:986)]
zygos_teo <- teosinte_snps[, c(1,3,5:986)]
melt_zygos_mz <- melt(zygos_mz, id.vars = c("Sample_ID","Group"),  variable.name = "SNP column", na.rm = FALSE, value.name = "SNPs", factorsAsStrings = TRUE)
melt_zygos_teo <- melt(zygos_teo, id.vars = c("Sample_ID", "Group"), variable.name = "SNP column", na.rm = FALSE, value.name = "SNPs", factorsAsStrings = TRUE)
```


Next I created new files to indicate whether a site was homozygous or heterozygous, as well as any missing data that is coded as NA. 

First I created the melted files:

```{r}
genotypes = colnames(fang)[-c(1:3)]
melt_geno = melt(fang, measure.vars = genotypes)
melt_geno[melt_geno == "?/?" ] = NA

```

Then I broke out the genotypes that were homozygous, heterozygous, or NA:

```{r}

SNPS_HoZ <- c("A/A", "C/C", "T/T", "G/G")
melt_geno$HoZ = (melt_geno$value=="A/A" | melt_geno$value=="C/C" | melt_geno$value=="T/T" | melt_geno$value=="G/G")
ordered_melt_geno <- melt_geno[, c(1,3,5,6)]

geno_ID = melt_geno[order(melt_geno$Sample_ID),]
geno_Group = melt_geno[order(melt_geno$Group), ]
Amt_per_ID = ddply(geno_ID, c("Sample_ID"), summarise, Homozygous=sum(HoZ, na.rm=TRUE), Heterozygous=sum(!HoZ, na.rm=TRUE), Missing=sum(is.na(HoZ)))
```

Then created the plot for the comparison of Heterozygosity, Homozygosity, and Missing
```{r}
melt_Amt_per_ID = melt(Amt_per_ID, measure.vars = c("Homozygous", "Heterozygous", "Missing"))
ggplot(melt_Amt_per_ID, aes(x = Sample_ID, y=value, fill = variable)) + geom_bar(stat = "identity", position = "stack")
```

#Part Two: Own Visualization

Next, I created new files to indicate whether a site was homozygous or heterozygous, as well as any missing data that is coded as NA, this time broken out by individual SNP combinations.

```{r}
SNPS_HoZ <- c("A/A", "C/C", "T/T", "G/G")
HoZyg_mz <- filter(melt_zygos_mz, SNPs== "A/A" | SNPs== "C/C" | SNPs== "G/G" | SNPs== "T/T")
HeZygC_mz <- filter(melt_zygos_mz, SNPs== "C/G" | SNPs== "C/A" | SNPs== "C/T")
HeZygG_mz <- filter(melt_zygos_mz, SNPs== "G/A" | SNPs== "G/C" | SNPs== "G/T")
HeZygA_mz <- filter(melt_zygos_mz, SNPs== "A/G" | SNPs== "A/C" | SNPs== "A/T")
HeZygT_mz <- filter(melt_zygos_mz, SNPs== "T/G" | SNPs== "T/A" | SNPs== "T/C")

Missing_mz <- replace(melt_zygos_mz, is.na(melt_zygos_mz), "N/A")
Missing_mz <- filter(melt_zygos_mz, SNPs== "N/A")

HoZyg_teo <- filter(melt_zygos_teo, SNPs== "A/A" | SNPs== "C/C" | SNPs== "G/G" | SNPs== "T/T")
HeZygC_teo <- filter(melt_zygos_teo, SNPs== "C/G" | SNPs== "C/A" | SNPs== "C/T")
HeZygG_teo <- filter(melt_zygos_teo, SNPs== "G/A" | SNPs== "G/C" | SNPs== "G/T")
HeZygA_teo <- filter(melt_zygos_teo, SNPs== "A/G" | SNPs== "A/C" | SNPs== "A/T")
HeZygT_teo <- filter(melt_zygos_teo, SNPs== "T/G" | SNPs== "T/A" | SNPs== "T/C")

Missing_teo <- replace(melt_zygos_teo, is.na(melt_zygos_teo), "N/A")
Missing_teo <- filter(melt_zygos_teo, SNPs== "N/A")

```

I used rbind() to combine all the rows into a single file. 

```{r}
Tally_HoZyg_mz <- HoZyg_mz %>% count(SNPs)
Tally_HeZygA_mz <- HeZygA_mz %>% count(SNPs)
Tally_HeZygC_mz <- HeZygC_mz %>% count(SNPs)
Tally_HeZygG_mz <- HeZygG_mz %>% count(SNPs)
Tally_HeZygT_mz <- HeZygT_mz %>% count(SNPs)
Tally_Missing_mz <- Missing_mz %>% count(SNPs)

Zygosity_Maize <- rbind(Tally_HoZyg_mz,Tally_HeZygA_mz,Tally_HeZygC_mz,Tally_HeZygG_mz,Tally_HeZygT_mz, Tally_Missing_mz)

Tally_HoZyg_teo <- HoZyg_teo %>% count(SNPs)
Tally_HeZygA_teo <- HeZygA_teo %>% count(SNPs)
Tally_HeZygC_teo <- HeZygC_teo %>% count(SNPs)
Tally_HeZygG_teo <- HeZygG_teo %>% count(SNPs)
Tally_HeZygT_teo <- HeZygT_teo %>% count(SNPs)
Tally_Missing_teo <- Missing_teo %>% count(SNPs)

Zygosity_Teosinte <- rbind(Tally_HoZyg_teo,Tally_HeZygA_teo,Tally_HeZygC_teo,Tally_HeZygG_teo,Tally_HeZygT_teo, Tally_Missing_teo)
```

I then created the plot for the Maize Zygosity Counts:

```{r}

colnames(Zygosity_Maize) <- c("Maize_SNPs", "Maize_SNP_counts")
colnames(Zygosity_Teosinte) <- c("Teosinte_SNPs", "Teosinte_SNP_counts")

ggplot(Zygosity_Maize) + geom_col(aes(x=Maize_SNPs, y=Maize_SNP_counts))

```

And the Teosinte Zygosity Counts:

```{r}
ggplot(Zygosity_Teosinte) + geom_col(aes(x=Teosinte_SNPs, y=Teosinte_SNP_counts))
```

 
