---
title: "R Notebook - Assignment"
output: html_notebook
---

This is the R Notebook for the R assignment. 

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).


START INFORMATION FROM HERE:

First I downloaded the files needed for the assignment:

```{r}
fang <- read.table("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-fall2017/master/UNIX_Assignment/fang_et_al_genotypes.txt", sep = "\t", header = TRUE, na.strings = "?/?", stringsAsFactors = FALSE)
```
```{r}
snp_pos <- read.table("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2017/master/UNIX_Assignment/snp_position.txt", sep = "\t", header = TRUE, na.strings = "?/?", stringsAsFactors = FALSE)
```

This allowed me to download the file, indicate that the files were tab delimited, there are headers in the first row, replaced any ?/? with NA, and prevented the automatic conversion of character strings to factors

Next, I viewed the data to take a quick look that all the data had downloaded correctly:

```{r}
View(fang)
View(snp_pos)
```

I then examine the structure and dimension of the fang et al files:
```{r}
typeof(fang)
class(fang)
ncol(fang)
nrow(fang)

```
I examine the structure and dimensions of the snp position files:

```{r}
typeof(snp_pos)
class(snp_pos)
ncol(snp_pos)
nrow(snp_pos)
```

I filter out the maize files from the fang et al file, first by creating a vector with the maize group IDs, then pulling the rows with those IDs into a new file. I then view the files to verify:

```{r}
maize_snps <- filter(fang, Group== "ZMMIL" | Group=="ZMMLR" | Group== "ZMMMR")
View(maize_snps)
```

I also filter out the teosinte files with the same process. First by creating a vector with the group IDs, then pulling the rows with those IDs into a new file, followed by a view to verify:
```{r}
teosinte_snps <- filter(fang, Group== "ZMPBA" | Group== "ZMPIL" | Group== "ZMPJA")
View(teosinte_snps)
```

Then I transposed the genotype data in order to join the two dataframes:
```{r}
transp_maize_snps <- t(maize_snps)
transp_teo_snps <- t(teosinte_snps)
```
I then review the transposed data:
```{r}
View(transp_maize_snps)
View(transp_teo_snps)
typeof(transp_maize_snps)
class(transp_maize_snps)
typeof(transp_teo_snps)
class(transp_teo_snps)
```

The transposed data vector was converted to characters/matrix, when as listed above, it should be list/data.frame. Therefore, I converted the transposed data into a dataframe, with a list type:

```{r}
df_trns_maize_snps <- as.data.frame(transp_maize_snps, stringsAsFactors = FALSE)
df_trns_teo_snps <- as.data.frame(transp_teo_snps, stringsAsFactors = FALSE)
```
I then reviewed the dataframe'd transposed data:

```{r}
View(df_trns_maize_snps)
View(df_trns_teo_snps)

typeof(df_trns_maize_snps)
class(df_trns_maize_snps)
typeof(df_trns_teo_snps)
class(df_trns_teo_snps)
```
With the values in the correct vector, I then merged the data frames together

```{r}
merged_mz_data <- merge.data.frame(snp_pos, df_trns_maize_snps, by.x = 1, by.y = 0)
merged_teo_data <- merge.data.frame(snp_pos, df_trns_teo_snps, by.x = 1, by.y = 0)
View(merged_mz_data)
View(merged_teo_data)
ncol(merged_mz_data)
ncol(merged_teo_data)
typeof(merged_mz_data)
class(merged_mz_data)
```


Now that the data has been merged, the files need to be ordered in the following order:

Column 1: SNP_ID  (currently column 1)
Column 2: Chromosome (currently column 3)
Column 3: Position (currently column 4)
Subsq col: all other genotype data (columns )

So I reordered the columsn in the following order:

maize: 1, 3, 4, 16-1588
teosinte: 1, 3, 4, 16-990

Omitting columns 2 and 5-15 as non-genotype data and thus not needed.

```{r}
ord_mz_data <- merged_mz_data[,c(1,3,4,16:1588)]
ord_teo_data <- merged_teo_data[, c(1,3,4,16:990)]
View(ord_mz_data)
View(ord_teo_data)
```
```{r}
typeof(ord_mz_data)
class(ord_mz_data)
```
I reviewed the type and class of both Position and Chromosome and found they were both Character:

```{r}
typeof(ord_mz_data$Position)
class(ord_mz_data$Position)
typeof(ord_mz_data$Chromosome)
class(ord_mz_data$Chromosome)
```
Therefore, I needed to convert these to doubles:

```{r}
ord_mz_data$Position <- as.factor(ord_mz_data$Position)
ord_mz_data$Chromosome <- as.factor(ord_mz_data$Chromosome)
ord_teo_data$Position <- as.factor(ord_teo_data$Position)
ord_teo_data$Chromosome <- as.factor(ord_teo_data$Chromosome)
```
```{r}
typeof(ord_mz_data$Position)
typeof(ord_mz_data$Chromosome)
typeof(ord_teo_data$Position)
typeof(ord_teo_data$Chromosome)
```


So I'm going to attempt to order them in increasing position with the missing data coded as "?":

```{r}
ord_mz_data_inc <- arrange(ord_mz_data, Position)
ord_teo_data_inc <- arrange(ord_teo_data, Position)
ord_mz_data_dec <- arrange(ord_mz_data, desc(Position))
ord_teo_data_dec <- arrange(ord_teo_data, desc(Position))
View(ord_mz_data_inc)
View(ord_teo_data_inc)
View(ord_mz_data_dec)
View(ord_teo_data_dec)
```
```{r}
typeof(ord_mz_data_dec$Chromosome)
typeof(ord_mz_data_dec$Position)
```

